src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -351,19 +315,90 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
         #region UrlEncode public methods
 
         [SuppressMessage("Microsoft.Design", "CA1055:UriReturnValuesShouldNotBeStrings", Justification = "Already shipped public API; code moved here as part of API consolidation")]
-        public static string UrlEncode(string value)
+        public static unsafe string UrlEncode(string value)
         {
             if (string.IsNullOrEmpty(value))
+            {
                 return value;
+            }
+            
+            bool foundSpaces = false;
+            int unsafeCount = 0;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            // count them first
+            var encoding = Encoding.UTF8;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+
+                if (ch == ' ')
+                {
+                    foundSpaces = true;
+                }
+                else if (!IsUrlSafeChar(ch))
+                {
+                    if (ch < 128) // optimize for ASCII chars
+                    {
+                        unsafeCount++; // UTF-8 uses 1 byte for chars 0 to 127
+                    }
+                    else if (char.IsSurrogatePair(value, i)) // this method should handle the case where i is the last index
+                    {
+                        // surrogate code points are encoded differently separate than when alone, so we need to address that
+                        i++;
+                        unsafeCount += encoding.GetByteCount(&ch, 2);
```

Comment:
Small optimization: this could be hard-coded as `4` instead (of course, it would have to be commented). See: http://ideone.com/vkpG92